programs/mango-v4/src/allocator.rs

```diff
@@ -0,0 +1,76 @@
+#![allow(dead_code)]
+
+use std::alloc::{GlobalAlloc, Layout};
+
+#[cfg(not(feature = "no-entrypoint"))]
+#[global_allocator]
+pub static ALLOCATOR: BumpAllocator = BumpAllocator {
+    start: solana_program::entrypoint::HEAP_START_ADDRESS as usize,
+};
+
+pub fn heap_used() -> usize {
+    #[cfg(not(feature = "no-entrypoint"))]
+    return ALLOCATOR.used();
+
+    #[cfg(feature = "no-entrypoint")]
+    return 0;
+}
+
+/// Custom bump allocator for on-chain operations
+///
+/// The default allocator is also a bump one, but grows from a fixed
+/// HEAP_START + 32kb downwards and has no way of making use of extra
+/// heap space requested for the transaction.
+///
+/// This implementation starts at HEAP_START and grows upward, producing
+/// a segfault once out of available heap memory.
+pub struct BumpAllocator {
+    pub start: usize,
+}
+
+unsafe impl GlobalAlloc for BumpAllocator {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let pos_ptr = self.start as *mut usize;
+
+        let mut pos = *pos_ptr;
+        if pos == 0 {
+            // First time, set starting position
+            pos = self.start + 8;
+        }
+
+        // Align pos to create the address for the allocation
+        let a = layout.align().wrapping_sub(1);
+        let mut begin = pos.checked_add(a).unwrap();
+        begin &= !a;
```

Comment:
bard to follow

Reply:
I'll add another comment. It's how to find the next `a`-aligned address `>=pos`